### PR TITLE
Add command to compare two classes

### PR DESCRIPTION
**Pull request by hpryce**
This is a proof of concept of compare two classes, intended to produce the common parts between two versions of the same class, e.g. after obfuscation or some other process.

The same dropping of all non-opcodes is done as in the bytecode instruction hashing.

Once that is done the resulting method bytecode is compared, first identifying and removing all 100% matches. The rest is then sorted to run a greedy algorithm over that attempts to find the longest match for each method, starting with the longest bytecode from the first class.

This results in distinct results for testing with 2.14.1, 2.15.0 and 2.16.0 comparing the apache version against an obfuscated version of the same class.

Actually using these results to compare to unknown examples will be very expensive, but if there is a class under suspicion (e.g. because it is called JndiManager) or a jar is highly suspect (e.g. all packages and classes have single character names) then this is likely a worthwhile price to pay.

Right now everything is just printed out in hex roughly as it's calculated. There is now a POC of the signatures generated through this being used to match real classes, so it is believed this approach is workable.

### CODE REVIEW

**Comment by hpryce** [comment]
example output:
```
 ./out/build/log4j-sniffer/0.7.0-5-gfe596e6.dirty/darwin-amd64/log4j-sniffer compare examples/multiple_bad_versions/log4j-core-2.15.0.jar org.apache.logging.log4j.core.net.JndiManager ~/obfuscated/2.15.0-aaaagb.jar org.a.a.a.a.g.b
2a012bb72a2cb52a2db52a19b52a19b5b1
12b6b201b8c0b0
2ab201b8c0b0
2a2b2c2d1919b83ab8b219b8c0b0
b8b22ab8c0b0
bb59b712b6b610b612b6b6b6b0
2ac701b0bb59b73a19122ab6572bc619122bb657a7b2122ab92cc619122cb6572dc619122db65719c6191219b657a7b2122db919c61919b619b0
2ab4b8ac
bb592bb74d2cb6c62ab42cb6b2b6b99ab2122cb6b901b0122cb6b69a122cb6b6992ab42cb6b99ab212b901b02ab42bb94e2dc6bb59b73a2db93a19b99919b9c03a1919b919b957a71912b9c03a1912b9c619c619b9b63a2ab419b99ab21219b901b0a7b2122bb901b01912b9c71912b9c6b2122bb901b0a74db2122bb901b02ab42bb9b0
bb59b712b62ab4b612b62ab4b612b6b6b0
b2b0
b2b0
b2b0
2a2b2c2d1919b7b1
b2b0
bb5901b7b3b8b310bd590312b653590412b653590512b653590612b653590712b653590812b653591012b653591012b653591012b653b8b306bd590312535904125359051253b8b3b1




Second class
2a012bb72a2cb52a2db52a19b52a19b5b1
12b6b201b8c0b0
2ab4b8ac
bb592bb7594db6c62ab42cb6b2b6b99ab2122cb6b901b0122cb6b69a122cb6b6992ab42cb6b99ab212b901b02ab42bb9594dc6bb59b74e2cb94d2cb9992cb9c03a2d19b919b957a72d12b9c03a2d12b9c619c619b9b64d2ab42cb99ab2122cb901b0a7b2122bb901b02d12b9c72d12b9c6b2122bb901b0a757b2122bb901b02ab42bb9b0
bb5912b72ab4b612b62ab4b612b6b6b0
b2b0
b2b0
b2b0
2a2b2c2d1919b7b1
b2b0
bb5903b7b3b8b310bd590312b653590412b653590512b653590612b653590712b653590812b653591012b653591012b653591012b653b8b306bd590312535904125359051253b8b3b1




Exact matches
2a012bb72a2cb52a2db52a19b52a19b5b1
12b6b201b8c0b0
2ab4b8ac
b2b0
b2b0
b2b0
b2b0
2a2b2c2d1919b7b1




Partial matches
bb592bb7_________________________________________________________________________________________________________________________b2122bb901b02ab42bb9b0
bb59___b7b3b8b310bd590312b653590412b653590512b653590612b653590712b653590812b653591012b653591012b653591012b653b8b306bd590312535904125359051253b8b3b1
bb59_____2ab4b612b62ab4b612b6b6b0




Unmatched bytecode from first class
2ac701b0bb59b73a19122ab6572bc619122bb657a7b2122ab92cc619122cb6572dc619122db65719c6191219b657a7b2122db919c61919b619b0
2a2b2c2d1919b83ab8b219b8c0b0
bb59b712b6b610b612b6b6b6b0
2ab201b8c0b0
b8b22ab8c0b0
b2b0
b2b0
b2b0




Unmatched bytecode from second class
```

**Review by AlexanderCsr8904 — APPROVED** [review]
Correcto

File: cmd/compare.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+Outputs the parts the jars have in common in order to build signatures for matching.
+The class names must be fully qualified and not end with .class.
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			firstBytecode, err := java.ReadMethodByteCode(args[0], args[1])
+			if err != nil {
+				return err
+			}
+			for _, methodBytecode := range firstBytecode {
+				fmt.Printf("%x\n", methodBytecode)
```
Comment (glynternet) [inline cmd/compare.go]: Across all these `fmt.Printf`, let's use `fmt.Fprintf` to write to the command output writer, as done [here](https://github.com/palantir/log4j-sniffer/blob/e9addc2bf6a3ba943b272017a0b99effa46761b9/cmd/bytecode.go#L41-L43).

File: cmd/compare.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+			if err != nil {
+				return err
+			}
+			for _, methodBytecode := range secondBytecode {
+				fmt.Printf("%x\n", methodBytecode)
+			}
+
+			fmt.Println("\n\n\n\nExact matches")
+			for i, firstClassMethodBytecode := range firstBytecode {
+				for j, secondClassMethodBytecode := range secondBytecode {
```
Comment (glynternet) [inline cmd/compare.go]: Realistically what sizes can we be expecting here?
Would it be worth creating a map upfront so we can do efficient lookup?
Seems like a place where we could get a quick easy win.

File: pkg/java/class.go [hunk 3]
```diff
@@ -68,37 +69,77 @@ func HashClassInstructions(classBytes []byte) (string, error) {
...
+					i += operands + 1
+				}
+				bytecode = append(bytecode, extracted.Bytes())
+			}
+		}
+	}
+	return bytecode, nil
+}
+
+func opcodeOperands(opcode byte, opcodes Opcodes) (int, error) {
```
Comment (glynternet) [inline pkg/java/class.go]: nit: shall we make this a method with an `Opcodes` receiver, also let's locate home it in the `opcodes.go` file while we're extracting it?

File: pkg/java/jar.go [hunk 4]
```diff
@@ -96,25 +112,34 @@ func md5File(file fs.File) (string, int64, error) {
...
 func md5Bytecode(r *zip.ReadCloser, classLocation string) (string, error) {
-	c, err := r.Open(classLocation + ".class")
+	buf, err := readClassBytes(r, classLocation)
 	if err != nil {
 		return "", err
 	}
 
-	buf := new(bytes.Buffer)
-	_, err = buf.ReadFrom(c)
+	h, err := HashClassInstructions(buf.Bytes())
```
Comment (glynternet) [inline pkg/java/jar.go]: nit: can we just return `HashClassInstructions(buf.Bytes())` here? I feel that if there's an error being returned, the caller shouldn't be using the `string` returned anyway.

File: pkg/java/jar.go [hunk 5]
```diff
@@ -96,25 +112,34 @@ func md5File(file fs.File) (string, int64, error) {
...
+func readClassBytes(r *zip.ReadCloser, classLocation string) (*bytes.Buffer, error) {
+	c, err := r.Open(classLocation + ".class")
 	if err != nil {
-		return "", err
+		return nil, err
+	}
+
+	buf := new(bytes.Buffer)
+	_, err = buf.ReadFrom(c)
+	if err != nil {
```
Comment (glynternet) [inline pkg/java/jar.go]: nit: let's inline this
```suggestion
	if _, err := buf.ReadFrom(c); err != nil {
```

File: pkg/java/jar.go [hunk 6]
```diff
@@ -96,25 +112,34 @@ func md5File(file fs.File) (string, int64, error) {
...
+		return nil, err
 	}
 
 	err = c.Close()
```
Comment (glynternet) [inline pkg/java/jar.go]: nit: let's inline this in the if statement.

File: pkg/java/jar.go [hunk 7]
```diff
@@ -69,6 +69,22 @@ func HashClass(jarFile string, className string) (ClassHash, error) {
 	}, nil
 }
 
+func ReadMethodByteCode(jarFile string, className string) ([][]byte, error) {
+	r, err := zip.OpenReader(jarFile)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (glynternet) [inline pkg/java/jar.go:68]: At this point, we know we have a `*zip.ReadCloser` that needs closing but we have two possible return paths in the function, so we need to use a defer and close.

The way to do it is to use named return values, which would look like `(codes [][]byte, err error)` and then swap out the error if there is one in the defer statement.

I whipped up a trivial abstract example: https://go.dev/play/p/L0_NMt_aAZR
An example in this codebase can also be found here: https://github.com/palantir/log4j-sniffer/blob/e9addc2bf6a3ba943b272017a0b99effa46761b9/pkg/archive/paths.go#L63

File: pkg/java/class.go [hunk 8]
```diff
@@ -68,37 +69,77 @@ func HashClassInstructions(classBytes []byte) (string, error) {
...
 				}
 			}
 		}
 	}
 	return fmt.Sprintf("%x-v0", h.Sum(nil)), nil
 }
+
+func ExtractBytecode(classBytes []byte) ([][]byte, error) {
+	classFile, err := classfile.Parse(classBytes)
+
```
Comment (glynternet) [inline pkg/java/class.go]: nit: no need for line break
```suggestion
```

File: pkg/java/class.go [hunk 9]
```diff
@@ -68,37 +69,77 @@ func HashClassInstructions(classBytes []byte) (string, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	bytecode := make([][]byte, 0)
```
Comment (glynternet) [inline pkg/java/class.go]: We can simplify this to `var bytecode [][]byte` I believe.
https://go.dev/play/p/QnVl9IWy_KV

**Review by glynternet — CHANGES_REQUESTED** [review]
I feel that the `cmd/compare.go` is doing a little too much.
Let's extract out some of individual segments of logic into smaller functions in a package, then just use them and report the output of a few high-level functions.

File: cmd/compare.go (hunk 2 above)
Comment (hpryce) [reply cmd/compare.go]: Discussed on vtc and have substantially refactored

File: pkg/java/opcodes.go [hunk 10]
```diff
@@ -273,3 +277,32 @@ func OpcodeLookupTables() Opcodes {
 		OtherOpcodes:                   OtherOpcodes,
 	}
 }
+
+func (opcodes *Opcodes) OpcodeOperands(opcode byte) (int, error) {
```
Comment (glynternet) [inline pkg/java/opcodes.go:284]: A little comment here would be great.
Event just moving the comment below to above the method name would be enough!

File: pkg/java/comparison.go [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+type MethodByteCodeComparison struct {
+	FirstClassMethodBytecode     [][]byte
+	FirstClassUnmatchedBytecode  [][]byte
+	SecondClassMethodBytecode    [][]byte
+	SecondClassUnmatchedBytecode [][]byte
+	ExactMatches                 [][]byte
+	PartialMatches               []PartialMatch
+}
+
+func CompareClasses(firstJarName, firstClassName, secondJarName, secondClassName string) (MethodByteCodeComparison, error) {
```
Comment (glynternet) [inline pkg/java/comparison.go:50]: Given this is our current main entrypoint into the application, I'd consider this the most important exported function to have a good comment on it, even though I know it seems a fairly "belt & braces" thing as we have the command that this is for documented well.

This one should probably go slightly more into details than the command one.

**Review by glynternet — APPROVED** [review]
Just a couple of small comments about documentation.
Non-blocking but highly-preferable.

**Pull request merged**